### ISSUE TITLE
We're having some performance issues.

### ISSUE THREAD

**Issue by victropolis** [brettwooldridge/HikariCP#1418]
Java version:

```
java -version
openjdk version "1.8.0_181"
OpenJDK Runtime Environment (build 1.8.0_181-b13)
OpenJDK 64-Bit Server VM (build 25.181-b13, mixed mode)
```

Postgres version:

```
dgcs=> select version();
                                                    version                                                     
----------------------------------------------------------------------------------------------------------------
 PostgreSQL 9.3.12 on x86_64-unknown-linux-gnu, compiled by gcc (GCC) 4.4.7 20120313 (Red Hat 4.4.7-16), 64-bit
(1 row)
```

Postgres Driver version:

```
            <dependency>
                <groupId>org.postgresql</groupId>
                <artifactId>postgresql</artifactId>
                <version>42.2.5</version>
            </dependency>
```

HikariCP version:

```
            <dependency>
                <groupId>com.zaxxer</groupId>
                <artifactId>HikariCP</artifactId>
                <version>3.3.1</version>
            </dependency>
```

HikariCP knobs:

```
        <maxLifetime>1800000</maxLifetime>
        <idleTimeout>60000</idleTimeout>
        <mininumPoolSize>1</mininumPoolSize>
        <maximumPoolSize>25</maximumPoolSize>
        <connectionTimeout>30000</connectionTimeout>
        <validationTimeout>5000</validationTimeout>
        <leakDetectionThreshold>5000</leakDetectionThreshold>
```

jdbcUrl:

```
<jdbcUrl>jdbc:postgresql://${GLOBAL.MEMBER_DB_SERVER}:${GLOBAL.MEMBER_DB_PORT}/${GLOBAL.MEMBER_DB_DATABASE}?socketTimeout=3600${GLOBAL.MEMBER_DB_SSL_URL_PARAMETER}</jdbcUrl>
```

We're having some performance issues.  Turns out we have hundreds (200+) of threads waiting at the following location and I'm trying to troubleshoot.  Any thoughts would greatly appreciated:

```
"http-bio-8082-exec-2084" #4946 daemon prio=5 os_prio=0 tid=0x00007f729c0f6800 nid=0x1e0a waiting on condition [0x00007f72376c6000]
   java.lang.Thread.State: TIMED_WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <merged>(a java.util.concurrent.SynchronousQueue$TransferQueue)
	at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:215)
	at java.util.concurrent.SynchronousQueue$TransferQueue.awaitFulfill(SynchronousQueue.java:764)
	at java.util.concurrent.SynchronousQueue$TransferQueue.transfer(SynchronousQueue.java:695)
	at java.util.concurrent.SynchronousQueue.poll(SynchronousQueue.java:941)
	at com.zaxxer.hikari.util.ConcurrentBag.borrow(ConcurrentBag.java:157)
	at com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:179)
	at com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:161)
	at com.zaxxer.hikari.HikariDataSource.getConnection(HikariDataSource.java:100)
	at org.springframework.orm.hibernate3.LocalDataSourceConnectionProvider.getConnection(LocalDataSourceConnectionProvider.java:87)
	at org.hibernate.jdbc.ConnectionManager.openConnection(ConnectionManager.java:446)
	at org.hibernate.jdbc.ConnectionManager.getConnection(ConnectionManager.java:167)
	at org.hibernate.jdbc.JDBCContext.connection(JDBCContext.java:160)
	at org.hibernate.transaction.JDBCTransaction.begin(JDBCTransaction.java:81)
	at org.hibernate.impl.SessionImpl.beginTransaction(SessionImpl.java:1473)
	at sun.reflect.GeneratedMethodAccessor281.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.hibernate.context.ThreadLocalSessionContext$TransactionProtectionWrapper.invoke(ThreadLocalSessionContext.java:344)
	at com.sun.proxy.$Proxy101.beginTransaction(Unknown Source)
```

Maybe we should lower `socketTimeout` to identify the long running queries?  Kinda makes sense that `socketTimeout` should be less than `connectionTimeout`.  Thoughts?

Thanks in advance!

**Comment by victropolis** [comment]
We did upgrade from HikaciCP `<version>2.4.4</version>` to `<version>3.3.1</version>` for this release of our  software.  I am currently trying to ascertain if anything else relevant may have changed.

**Comment by victropolis** [comment]
Changing the postgres `socketTimeout` is having no effect.  Could HikariCP be doing something to override that?  Thanks!

**Comment by victropolis** [comment]
@brettwooldridge UPDATE: Tried rolling back HikariCP to version 2.4.4 and initial indication are that this was helpful.  Going back to 3.3.1 to confirm that upgrading reproduces the problem described above.

**Comment by victropolis** [comment]
```
"http-bio-8080-exec-200" #292 daemon prio=5 os_prio=0 tid=0x00007f6624194000 nid=0x4f02 waiting on condition [0x00007f65aac96000]
   java.lang.Thread.State: TIMED_WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <merged>(a java.util.concurrent.SynchronousQueue$TransferQueue)
	at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:215)
	at java.util.concurrent.SynchronousQueue$TransferQueue.awaitFulfill(SynchronousQueue.java:764)
	at java.util.concurrent.SynchronousQueue$TransferQueue.transfer(SynchronousQueue.java:695)
	at java.util.concurrent.SynchronousQueue.poll(SynchronousQueue.java:941)
	at com.zaxxer.hikari.util.ConcurrentBag.borrow(ConcurrentBag.java:157)
	at com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:179)
	at com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:161)
	at com.zaxxer.hikari.HikariDataSource.getConnection(HikariDataSource.java:100)
```

Still inconclusive.  under 3.3.1 running on my local box, _sometimes_ I see a lot of threads waiting there which is consistent with the thread dumps I've seen from proooduction.  I have seen fewer situations under 2.4.4.  Continuing to test.

**Comment by victropolis** [comment]
The waiting threads may be a red herring.  the tomcat access logs don't seem to indicate that the service in question is the bottleneck.  Stand by.

**Comment by brettwooldridge** [comment]
@victropolis Threads get to this point...
```
com.zaxxer.hikari.util.ConcurrentBag.borrow(ConcurrentBag.java:157)
```
when there are no connections available in the pool.  Basically, the thread is waiting in ``borrow()`` for a direct hand-off of a connection returning to the pool, which occurs when a connection is "closed".

Lot's of threads parked in ``borrow()`` is not necessarily a problem if their time spent there is short; on the order of milliseconds.  However, if threads are parked there for more than a few seconds, that indicates to me that queries are taking longer than expected, or connections are being held out of the pool for longer than necessary.

One thing you might do is examine query performance from the server-side by instructing PostgreSQL to collect query statistics.  Another approach is to setup metrics, either Dropwizard, micrometer, or Prometheus -- from which you can observe:
 * Connection acquisition time, and
 * Connection usage (out-of-the-pool) time

Both should be extremely short, IMHO, well under a second.

**Comment by JoyceLikeRose** [comment]
Maybe transaction is too long or exists big loop in your program,
that hold all connections in using status with long time?

**Comment by victropolis** [comment]
@brettwooldridge that's all well and good, but why would simply downgrading back to 2.4.4 and changing nothing else produce fewer threads in such a state and upgrading to 3.3.1 produce more threads in such a state?  Anyway, we are currently pursuing other avenues, but may downgrade to 2.4.4 if we are unsuccessful.

**Comment by victropolis** [comment]
@brettwooldridge FALSE ALARM!  We found a problem in our code that was basically bypassing a caching mechanism and was resulting in millions more database calls.  HikariCP is solid!

**Comment by chaima-ennar** [comment]
@victropolis  hello ! we are facing the same issue, could you please explain more what is the bug in your app. Maybe we are doing the same without knowing it. Thanks in advance.

**Comment by victropolis** [comment]
@chaima-ennar I seriously doubt it.  We have a cache in place to reduce the number of database calls and someone put in a new database call that was completely bypassing the cache, which meant we had A LOT more database calls.  So, completely unrelated to HikariCP and totally our devs' fault.  Good luck!

**Comment by tbadalov** [comment]
> However, if threads are parked there for more than a few seconds, that indicates to me that queries are taking longer than expected, or connections are being held out of the pool for longer than necessary.

What looks strange to me is that it took ~8.5 seconds to throw a timeout exception with a message "Connection is not available, request timed out after **1128ms** as if the timeout was delayed.

Hikari 5.1.0
Hibernate 6.5.3
Spring Boot 3.9.0


connect-timeout: 1000

<img width="1403" alt="Image" src="https://github.com/user-attachments/assets/fd7a225b-0de9-467d-bed0-185e94988c3b" />

**Comment by tbadalov** [comment]
The app was CPU-throttled. This didn't happen after CPU limits were removed
